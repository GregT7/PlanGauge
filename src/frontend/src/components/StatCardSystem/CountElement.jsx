function CountElement({styling, count}) {
    return (
    <div className="flex items-center gap-1">
        <div className={`w-12 h-12 rounded-full ${styling}`}></div>
        <span><span>{`x${count}`}</span></span>
    </div>
)
}

export default CountElement;