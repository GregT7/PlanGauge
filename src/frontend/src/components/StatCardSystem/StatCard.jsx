import {
  Card,
  CardAction,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from "@/components/ui/card"
import toLocalMidnight from "@/utils/toLocalMidnight";
import { format } from "date-fns";

function capitalizeFirstLetter(str) {
  if (typeof str !== 'string' || str.length === 0) {
    return str; // Handle non-string or empty inputs
  }
  return str.charAt(0).toUpperCase() + str.slice(1);
}

function determineStatusStyle(status) {
    switch (status) {
        case "neutral":
            return "bg-zinc-900";
        case "good":
            return "bg-emerald-900 border-2 border-emerald-400"
        case "moderate":
            return "bg-amber-900 border-2 border-amber-400"
        case "poor":
            return "bg-rose-900 border-2 border-rose-400";
        default:
            return "bg-red-900 border-2 border-red-500"
    }
}

const StatCard = ({cardData, className=""}) => {
    const cardDate = toLocalMidnight(cardData.date);
    const desciptStr = format(cardDate, 'MMMM d, yyyy')
    
    return (
        <Card className={`${className} ${determineStatusStyle(cardData.status)}`}>
            <CardHeader>
                <CardTitle>{cardData.name}</CardTitle>
                <CardDescription>{desciptStr}</CardDescription>
            </CardHeader>
            <CardContent>
                <p>{`Status: ${capitalizeFirstLetter(cardData.status)}`}</p>
                <p>{`Sum: ${cardData.sum} mins`}</p>
            </CardContent>
        </Card>
    );
}

export default StatCard;