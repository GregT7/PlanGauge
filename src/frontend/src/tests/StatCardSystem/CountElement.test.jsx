import { expect, it, describe } from "vitest";
import { screen, render } from "@testing-library/react";
import StatCard from "@/components/StatCardSystem/StatCard";
import CountElement from "@/components/StatCardSystem/CountElement";

describe("CountElement unit tests", () => {
    it("passes snapshot test", () => {
        const styling = "bg-emerald-500";
        const count = 3;
        const { container } = render(<CountElement styling={styling} count={count}/>)
        expect(container).toMatchSnapshot();
    });

    it("handles empty 'count' and 'styling' props without errors", () => {
        const { container } = render(<CountElement/>)
        expect(container).toBeInTheDocument();
    });

    it("handles invalid 'count' and 'styling' props without errors", () => {
        const count = {'objKey': 123};
        const styling = 512;
        const { container } = render(<CountElement count={count} styling={styling}/>)
        expect(container).toBeInTheDocument();
    });

    // applies default styling and count when invalid props are passed
    it("", () => {
        const count = {'objKey': 123};
        const styling = 512;
        const { container } = render(<CountElement count={count} styling={styling}/>)
        
        expect(screen.getByText('x0')).toBeInTheDocument();
        expect(screen.getByTestId('circle-div')).toHaveClass('bg-stone-300')
    })

    it("applies correct styling and count passed as props", () => {
        const count = 5;
        const styling = "bg-teal-200";
        render(<CountElement count={count} styling={styling}/>)

        expect(screen.getByText("x5")).toBeInTheDocument();
        expect(screen.getByTestId('circle-div')).toHaveClass(styling);
    });


})