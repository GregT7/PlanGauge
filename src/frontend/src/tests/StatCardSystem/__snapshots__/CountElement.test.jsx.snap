// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CountElement unit tests > passes snapshot test 1`] = `
<div>
  <div
    class="flex items-center gap-1"
  >
    <div
      class="w-12 h-12 rounded-full bg-emerald-500"
      data-testid="circle-div"
    />
    <span>
      <span>
        x3
      </span>
    </span>
  </div>
</div>
`;
